### DISCUSSION TITLE
How to show the scriptkit autocomplete with only subset of the scripts

### DISCUSSION THREAD

**Discussion by nadeemkhedr** [johnlindquist/kit#419] [Q&A] [answered]
Usecase, I have a bunch of scripts related to vim, I want to only show them when I hit a shortcut 

I saw the `left-click.js` and tried to create a new script with the same content 

```javascript
let script = await arg(
  `Which script do you want to run?`,
  //Put "choices" inside of function when caching to avoid re-running
  async () => (await cli("scripts")).scripts,
  //Enable caching
  true
)

await run(script)
```

I assume I can filter on `cli('scripts')` but the issue is that line throw an exception it tries to import `scripts` file but can't find it

**Comment by johnlindquist** [comment, marked as answer]
I need to update `left-click` click...

## scripts db / getScripts

There's a json file at `~/.kit/db/scripts.json` which the app auto-updates when anything in the `~/.kenv/scripts` dir changes. You can get the contents like so:

```js
let scripts = await getScripts()

inspect(scripts)
```

## selectScript

There's a built-in helper for prompting for a script:

```js
import { selectScript } from "kit/utils.js"

let script = await selectScript()

inspect(script)
```

Here's an example of "only show scripts whose command starts with "t"`:

```js
import { selectScript } from "kit/utils.js"

let placeholder = "Start with T"
let fromCache = true //false would re-scan the ~/.kenv/scripts dir in case you manually change something

let scriptsTransform = scripts =>
  scripts.filter(script => script.command.startsWith("t"))

let script = await selectScript(
  placeholder,
  fromCache,
  scriptsTransform
)

inspect(script)
```

**Reply by nadeemkhedr** [reply]
@johnlindquist that works! Thanks

The only difference is if I'm in `scripts`, and I import `kit` it throws an error, it needs to be relative 

This fails
```
import { selectScript } from 'kit/utils.js'
```

but this works
```
import { selectScript } from '../kit/utils.js'
```

Am I missing a `tsconfig.json` or something like that to specify root dir for scripts or for scripts its normal?

**Reply by johnlindquist** [reply]
Please post the error log from the failure.

Also, please post `~/.kenv/package.json`

Sidenote: a `tsconfig.json` with alias/rootdir would only impact .ts file import paths before the compile step. Script Kit doesn't have a compile step so any imports will be impacted by the dependency definitions in the kenv `package.json` and/or the package.json's of the installed deps  (the "main" and "exports" fields).

**Reply by nadeemkhedr** [reply]
Sure

My `~.kenv/package.json`
```
{
  "type": "module"
}
```

error log: 
```
[2021-09-01 19:09:18.052] [warn]  Cannot find package 'kit' imported from /Users/nadeemkhedr/.kenv/scripts/show-github-scripts.js
[2021-09-01 19:09:18.053] [warn]  Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'kit' imported from /Users/nadeemkhedr/.kenv/scripts/show-github-scripts.js
    at new NodeError (node:internal/errors:329:5)
    at packageResolve (node:internal/modules/esm/resolve:714:9)
    at moduleResolve (node:internal/modules/esm/resolve:755:18)
    at Loader.defaultResolve [as _resolve] (node:internal/modules/esm/resolve:869:11)
    at Loader.resolve (node:internal/modules/esm/loader:86:40)
    at Loader.getModuleJob (node:internal/modules/esm/loader:230:28)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:57:40)
    at link (node:internal/modules/esm/module_job:56:36)
[2021-09-01 19:09:18.054] [warn]  ☠️ ERROR PROMPT SHOULD SHOW ☠️

```

**Reply by johnlindquist** [reply]
hmm, that should  at least be:

```js
{
  "type": "module",
  "dependencies": {
    "kit": "file:../.kit"
  }
}
```

Did you happen to install way back in March?
https://github.com/johnlindquist/kit/discussions/112#discussion-3305311

**Reply by nadeemkhedr** [reply]
Yea, my installation is pretty old. code completion would be so awesome, I don' have that :D will fix it, thanks again man

**Reply by nadeemkhedr** [reply]
Just reinstalled it, have a different issue when running your code, getting this error
```
Package subpath './utils.js' is not defined by "exports" in /Users/nadeemkhedr/.kenv/node_modules/kit/package.json imported from /Users/nadeemkhedr/.kenv/scripts/show-github-scripts.js
```

I checked the [package.json](https://github.com/johnlindquist/kit/blob/beta/package.json) for the kit repo, and when i add the utils as an export it works fine. Is it missing or something wrong is happening?

**Reply by johnlindquist** [reply]
Looks like my fault. I need to push an update to the package.json.

Fyi: You can leave your changes to the `~/.kit/package.json`, but just be aware that each time Kit.app updates, it will clear out all your changes.
